Add GTP primitives to artix7 architectures

acomodi opened the pull request:
This is a PR which adds GTP primitives to the SymbiFlow toolchain:

- [x] GTPE2_COMMON
- [x] IBUFDS_GTE2
- [x] GTPE2_CHANNEL

For artix7 achitectures:

- [x] xc7a50t
- [x] xc7a100t
- [x] xc7a200t

acomodi commented:
Apart from 200T vendor check, and install, all the other CIs went green.

I am not sure what went wrong in the failing ones:

- [install](https://source.cloud.google.com/results/invocations/118b30cd-615c-4288-9988-346951c95f88/targets): This one got killed and no output log was produced.
- [200T Vendor](https://source.cloud.google.com/results/invocations/277b1d8e-3c35-4e36-89cd-aa9c36b0c725/targets): Given that not all targets completed, and no error log was produced, this may correspond to some part of the flow being stuck somewhere preventing the completion of some of the targets.

litghost commented:
> * [install](https://source.cloud.google.com/results/invocations/118b30cd-615c-4288-9988-346951c95f88/targets): This one got killed and no output log was produced.

Maybe a deadlock issue?  Vivado doesn't get invoked in the `install` CI?

``` 09:59:22 up 19:55,  0 users,  load average: 0.01, 0.02, 0.00
              total        used        free      shared  buff/cache   available
Mem:           102G        856M         40G         16M         60G        100G
Swap:          1.0G         25M        998M
Filesystem      Size  Used Avail Use% Mounted on
udev             52G     0   52G   0% /dev
tmpfs            11G   17M   11G   1% /run
/dev/sda1        99G   69G   26G  73% /
tmpfs            52G     0   52G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs            52G     0   52G   0% /sys/fs/cgroup
/dev/sdc1       246G   43G  203G  18% /opt/Xilinx
cgmfs           100K     0  100K   0% /run/cgmanager/fs
/dev/sdb1       4.0T   86G  3.9T   3% /tmpfs
tmpfs            11G     0   11G   0% /run/user/1000
/dev/sdd        246G   29G  217G  12% /image
Sat Mar 20 10:04:22 PDT 2021
 10:04:22 up 20:00,  0 users,  load average: 0.00, 0.00, 0.00
              total        used        free      shared  buff/cache   available
Mem:           102G        856M         40G         16M         60G        100G
Swap:          1.0G         25M        998M
Filesystem      Size  Used Avail Use% Mounted on
udev             52G     0   52G   0% /dev
tmpfs            11G   17M   11G   1% /run
/dev/sda1        99G   69G   26G  73% /
tmpfs            52G     0   52G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs            52G     0   52G   0% /sys/fs/cgroup
/dev/sdc1       246G   43G  203G  18% /opt/Xilinx
cgmfs           100K     0  100K   0% /run/cgmanager/fs
/dev/sdb1       4.0T   86G  3.9T   3% /tmpfs
tmpfs            11G     0   11G   0% /run/user/1000
/dev/sdd        246G   29G  217G  12% /image


ERROR: Aborting VM command due to timeout of 72000 seconds```

litghost commented:
> * [install](https://source.cloud.google.com/results/invocations/118b30cd-615c-4288-9988-346951c95f88/targets): This one got killed and no output log was produced.


```Flow timing analysis took 0.00313282 seconds (0.000682797 STA, 0.00245003 slack) (1 full updates: 1 setup, 0 hold, 0 combined).
VPR succeeded
The entire flow of VPR took 9.47 seconds (max_rss 3281.7 MiB)
cd build_basys3 && symbiflow_place -e top.eblif -d xc7a50t_test -n top.net -P xc7a35tcpg236-1 -a "--seed 1024"
Using default VPR options.
Generating coinstrains ...
usage: prjxray_create_place_constraints.py [-h] [--input INPUT]
                                           [--output OUTPUT] --net NET
                                           --vpr_grid_map VPR_GRID_MAP --arch
                                           ARCH --db_root DB_ROOT --part PART
                                           --blif BLIF [--roi]
                                           [--allow-bufg-logic-sources]
                                           [--graph_limit GRAPH_LIMIT]
prjxray_create_place_constraints.py: error: the following arguments are required: --db_root, --part
make: *** [Makefile:25: build_basys3/top.place] Error 2


0% tests passed, 1 tests failed out of 1

Total Test time (real) =  29.66 sec

The following tests FAILED:
	  1 - binary_toolchain_test_50t (Failed)
Errors while running CTest


[ID: 3369515] Build finished after 16266 secs, exit value: 1
```

Likely a disk-space overflow on toolchain test failure.  Should probably fix that.

acomodi commented:
> Looks like Vivado got stuck on the 200T CI:

I wonder what this could be due to... no GTP test is currently performed in the diff_fasm flow, so "hopefully" this is an intermittent issue. I'll try to reproduce it locally

> Likely a disk-space overflow on toolchain test failure. Should probably fix that.

Ok, thanks for the info, I'll try to figure out what is happening here

acomodi commented:
I have added the XML autogeneration with timings from the SDF directory and modified the [autogeneration script]
(https://github.com/SymbiFlow/symbiflow-arch-defs/pull/2073). This requires https://github.com/SymbiFlow/prjxray/pull/1628 to have more precise information of the port types from the `<primitive>_ports.json`.

Given that the cells_sim and techmap require some manual modifications, those are not part of the flow at the moment, at least until a better generalization mechanism is added for those files.
An idea might be to split the quite large cells_sim and cells_map into the belonging primitives and merge it. This way we can have a more fine-grained approach that can let us auto-generate at least some of the primitives mapping files.

acomodi commented:
>> Likely a disk-space overflow on toolchain test failure. Should probably fix that.

> Ok, thanks for the info, I'll try to figure out what is happening here

This was actually due to a missing update to the toolchain wrappers, as the device info needs to be available now at placement constriants generation, to correctly constrain IOPADs of the GTP tiles at the correct sub_tile location. Fixed now

acomodi commented:
@litghost CI went green, this PR is ready for review. Given that the XML autogeneration requires https://github.com/SymbiFlow/prjxray/pull/1628, for now the XML are checked-in, and can either be removed later on, or we can wait for https://github.com/SymbiFlow/prjxray/pull/1628 to land and the output integrated in the db

litghost commented on xc/xc7/toolchain_wrappers/vpr_common:
```diff
@@ -76,6 +80,7 @@ function parse_args {
      fi
 
      export DEVICE=$DEVICE
+     export BITSTREAM_DEVICE=$BITSTREAM_DEVICE
```
Can you explain why we have a `DEVICE` and a `BITSTREAM_DEVICE`?  Seems redundant?

acomodi replied:
Yeah, I should definetely rename these, in the meantime:

- DEVICE = actual name of the device from symbiflow-arch-defs, e.g. `xc7a50t_test`, from which the RR graph and all other files are named after.
- BITSTREAM_DEVICE = xc7 family, e.g. artix7, zynq7. This is used to access the db.

litghost replied:
But isn't the information redundant?  Doesn't `xc7a50t_test` imply `artix7`?

acomodi replied:
Well, sure, we could get the place_constraints script (or the toolchain wrappers) to infer that `xc7a50t_test` corresponds to the `artix7` family, I didn't really want to rely on implications and have the artix7 explicitly added as an argument for safety.

litghost replied:
I guess what really doesn't make sense is why a prjxray specific variable is needed in the script called `vpr_common`?  I thought we were trying to create a unified set of scripts, and this seems like it is moving in the wrong direction?

acomodi replied:
So, for context, the problem here was that the place constraints for the IOPADs for the GTP need to respect the way we are building the GTP tiles, and these must be in sync, otherwise constrained sub_tiles won't match, causing routing issues.

It might be possible to store the information on how to build and on how to constrain the IOPADs in the vpr_grid.csv or some additional data file, so we do not need to access prjxray

litghost replied:
But nothing you just wrote has to do with VPR itself?  `vpr_common` is for handling invocation of VPR.  Handling placement constraints happens before invoke VPR, yes?

acomodi replied:
Well, as part of the toolchain wrappers, the `symbiflow_generate_constraints` part is invoked within the `symbiflow_place` wrapper, therefore the `bitstream_device` argument needed to be passed to the `symbiflow_place` command.

The generate_constraints step consists in calling the constraint generation scripts, prior to invoking the VPR placer.

 In reality, VPR  itself does not need this information at all.

litghost replied:
I believe that `symbiflow_generate_constraints` should deal with mapping `DEVICE` to `BITSTREAM_DEVICE`, and that we should not be adding more arguments to the top level wrapper. This detail appears to be prjxray specific, and that probably shouldn't leak into the broader scripts.

acomodi replied:
All right, I have added the mapping into the constraints generation script

litghost commented on xc/common/utils/prjxray_create_pinmap_csv.py:
```diff
@@ -66,13 +72,32 @@ def main():
+                continue
+
+            gridinfo = grid.gridinfo_at_tilename(line['tile'])
+            sites = list(gridinfo.sites.keys())
```
This probably needs to be `sorted(...)`, not `list(...)`?

acomodi replied:
Done

litghost commented on xc/common/utils/prjxray_arch_import.py:
```diff
@@ -526,12 +534,22 @@ def make_prefix(site, x, y, from_site_name=False):
-        return site_type, '{}_Y{}'.format(site_type, y % prefix_required[1])
+        mod_y = prefix_required[1]
+        y_formula = "y{}".format(" % mod_y") if mod_y else "y"
+        return site_type, '{}_Y{}'.format(site_type, eval(y_formula))
```
Please do not use eval.  I don't believe it is required here.

acomodi replied:
Yep, removed eval

litghost commented on xc/common/utils/prjxray_form_channels.py:
```diff
@@ -2159,6 +2107,155 @@ def create_vpr_grid(conn, grid_map_output):
     write_cur.execute("ANALYZE")
     write_cur.execute("""COMMIT TRANSACTION;""")
 
+    fieldnames = [
```
Can you add some comments through here?

acomodi replied:
Comment added

litghost commented on xc/common/utils/prjxray_import_tile_capacity.py:
```diff
@@ -35,41 +46,58 @@ def main():
+        sites.append((site_type, site, input_wires, output_wires))
 
-        sites[site.type].append((site, input_wires, output_wires))
+    # If enabled, sort sites by name
```
Is there ever a reason to not sort them?

acomodi replied:
Actually yes, I bumped into an issue after applying sorting to all the capacity, I got an error for which the SLICEM and SLICEL were swapped in the sub_tile definitions, causing the wrong prefix to be assigned, given the different placement.

Thinking about it, I might have missed to sort also when generating the architecture layout, therefore having the tiles and their location prefixes not in sync.

As a long term more robust solution though, we would need to enhance VPR so that the sub_tile index is not defined by the order of the sub tile, but by an attribute that we can control.

acomodi replied:
Fixed the tile prefix ordering in the arch-layout, so now all sites are sorted.

litghost commented on xc/common/utils/prjxray_import_tile_capacity.py:
```diff
@@ -35,41 +46,58 @@ def main():
-        sites[site.type].append((site, input_wires, output_wires))
+    # If enabled, sort sites by name
+    if args.sort_sites:
+        sites = sorted(sites, key=lambda tuple: tuple[1].type)
```
Should sort by `(site.type, site.name)`?

acomodi replied:
Sure, done

litghost commented on xc/common/utils/prjxray_form_channels.py:
```diff
@@ -2159,6 +2107,164 @@ def create_vpr_grid(conn, grid_map_output):
+            # and will be implemented as a sub_tile. This information is relevant
+            # to generate correct placement constraints for sub_tiles directly
+            # connected within the same physical tile
+            connected_to_site = sites_with_direct.get(site_name, None)
```
Rather than "None", should the default be ""?

acomodi replied:
Actually `None` gets translated as an empty string in the csv, and read back as empty strings

litghost commented on xc/common/utils/prjxray_create_pinmap_csv.py:
```diff
@@ -66,13 +72,34 @@ def main():
+            gridinfo = grid.gridinfo_at_tilename(line['tile'])
+            sites = sorted(gridinfo.sites.keys())
+
+            print(sites)
```
Probably don't want this print anymore?

acomodi replied:
Indeed, removed

acomodi commented:
@mithro @litghost  Seems like there are troubles with kokoro (maybe related to https://github.com/SymbiFlow/prjxray/issues/1578#issuecomment-809843919 somehow?)

mithro commented:
@acomodi - I'm updating the base image to fix the permission issue.

litghost commented on xc/common/cmake/project_ray.cmake:
```diff
@@ -662,15 +671,17 @@ function(PROJECT_RAY_TILE_CAPACITY)
+  # UNUSED_WIRES: optional list of wires in site to be dropped
 
-  set(options)
+  set(options SORT_SITES)
```
Unused option now, remove?

acomodi replied:
Sure, removed

acomodi commented:
>  I'm updating the base image to fix the permission issue.

@mithro seems to be working now, thanks. There seems to be yet another issue with the [vendor tool tests](https://source.cloud.google.com/results/invocations/34007d60-9dee-45fc-b68c-a95b6813544f/targets/foss-fpga-tools%2Fsymbiflow-arch-defs%2Fpresubmit%2Fxc7_vendor/log)

litghost commented on xc/common/utils/prjxray_create_pinmap_csv.py:
```diff
@@ -66,13 +72,32 @@ def main():
+                continue
+
+            gridinfo = grid.gridinfo_at_tilename(line['tile'])
+            sites = sorted(gridinfo.sites.keys())
```
You should be able to get this from the connection_database?

acomodi replied:
Done
